support/src/traits.rs

```diff
@@ -80,3 +85,50 @@ pub trait GetWorstCase {
 pub trait ItemFilter<Item> {
 	fn should_include(&self, credential: &Item) -> bool;
 }
+
+pub trait StorageItemMeter<AccountId, Key> {
+	type Currency: ReservableCurrency<AccountId>;
+
+	fn deposit(
+		key: &Key,
+	) -> Result<Deposit<AccountId, <Self::Currency as Currency<AccountId>>::Balance>, DispatchError>;
+
+	fn deposit_amount(key: &Key) -> <Self::Currency as Currency<AccountId>>::Balance;
+
+	fn store_deposit(
```

Comment:
Since we have knowledge of the `AccountId`, I think it would be cleaner if the `store_deposit` function would take the deposit owner, and does everything internally, returning the deposit in the result. Returning the new deposit could also be something that `change_deposit_owner` and `update_deposit` could do.

Reply:
the goal of the trait is to reduce code duplication. That's done by moving the parts that change to the pallet, while the parts that stay the same into the default trait implementation. The code that never changes are actually only three lines of code

```rust
free_deposit::<AccountId, Self::Currency>(&deposit);

let deposit = Deposit {
	owner: new_owner,
	..deposit
};
Self::Currency::reserve(&deposit.owner, deposit.amount)?;
```

If we would now move parts of this to the `store_deposit` function, we would duplicate code again. I guess with "cleaner" you mean that the deposit amount and reserve is set in the `store_deposit` call and therefor an invalid state is much harder to achieve?

Reply:
I mean that the caller of the trait does not have to have knowledge of how to build a deposit, since that would become responsibility of the trait implementer.
I guess I don't fully understand your argument about code duplication. Now you would have to create the deposit type outside and give it to the `store_deposit()` function every time a deposit is to be reserved. If it goes inside the `store_deposit()` function, it is only implemented once by the trait implementer, and the client could access it as the function return value or, if it does not care, could simply ignore the result. Does it make sense what I am saying?

Reply:
The `update_deposit()` could then simply call `free_deposit()` and then return whatever `store_deposit()` returns, aka the new deposit or web3name ownership stuff.

Reply:
I guess what I am saying is that while `store_deposit()` takes a deposit as a parameter, `update_deposit()` does everything internally. I think the latter is a better approach.

Reply:
I'm in favour of the current approach! I also don't see what we win by removing the deposit input arg from `store_deposit`. You would just move the creation of the deposit into `store_deposit` and thus create duplicate code for all implementations of the  `StorageDepositCollector` trait.

IMO, it's super clean as is! E.g., the implementing pallets only call `change_deposit_owner` and `update_deposit` which generically create the deposit and forward that to that custom implementations of `store_deposit`.